onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -52,7 +54,190 @@ static void BuildFusedKernelDef(KernelDefBuilder& builder, const IndexedSubGraph
+             modified;
+
+  return Status::OK();
+}
+
+static Status GetCapabilityForEP(Graph& graph, KernelRegistryManager& kernel_registry_mgr, IExecutionProvider& current_ep,
+                                 GraphPartitioner::Mode mode, std::vector<std::unique_ptr<ComputeCapability>>& capabilities,
+                                 const logging::Logger& logger) {
+  {
+    GraphViewer graph_viewer(graph);
+    capabilities = current_ep.GetCapability(graph_viewer, kernel_registry_mgr.GetKernelRegistriesByProviderType(current_ep.Type()));
+  }
+
 #if !defined(ORT_MINIMAL_BUILD)
+  // Run layout transformer only for EPs other than CPU EP and provided the preferred layout is NHWC
+  // CPU EP layout transformation happens later when level 3 transformers are run.
+  if (mode != GraphPartitioner::Mode::kAssignOnly &&
+      current_ep.GetPreferredLayout() == DataLayout::NHWC) {
+    for (auto& capability : capabilities) {
+      // in theory an EP could return an empty value...
+      if (!capability->sub_graph) {
+        continue;
+      }
+      AssignNodes(graph, *capability->sub_graph, current_ep.Type());
+    }
+
+    // Perform layout transformation on the specific EP assigned graph
+    bool modified = false;
+    ORT_RETURN_IF_ERROR(TransformLayout(graph, modified, current_ep, logger));
+
+    // It is possible some new nodes are introduced during transformation. These nodes can be either existing nodes
+    // which are reconstructed to update domain or completly new nodes which are necessary for layout transformation.
+    // Therefore, we re-run GetCapability so that these new nodes can be processed by this EP.
+    if (modified) {
+      capabilities.clear();
+      GraphViewer graph_viewer(graph);
+      capabilities = current_ep.GetCapability(graph_viewer, kernel_registry_mgr.GetKernelRegistriesByProviderType(current_ep.Type()));
+    }
+  }
+#endif  // !defined(ORT_MINIMAL_BUILD)
+
+  return Status::OK();
+}
+
+/// <summary>
+/// Validate all the layout sensitive nodes which were transformed for current EP are indeed taken by current EP.
+/// If not, then we have a bug. If a node with domain kMSNHWC is left in the graph at this point then
+/// graph.Resolve will fail.
+/// Since layout transformation is only enabled for compile based EPs, just checking that graph does not contain
+/// a node with kMSNHWC domain is enough. This is because after compile all the nodes which the EP claims are fused
+/// into 1 and removed from the graph.
+/// </summary>
+/// <param name="graph">Graph to validate</param>
+/// <returns></returns>
+static Status ValidateGraphPartitioning(const Graph& graph) {
+  for (const auto& node : graph.Nodes()) {
+    if (node.Domain() == kMSNHWCDomain) {
+      return Status(common::ONNXRUNTIME, common::FAIL,
+                    "Graph contains an invalid node: " + node.Name() + " Op Type: " + node.OpType() +
+                        " with domain com.microsoft.nhwc. This is a bug in layout transformer.");
```

Comment:
Updated the comment to reflect it is a temporary node. I decided to keep the domain info in the meesage because it will be extremely easy for us to pin point where to start debugging.

If you see a node with kMSNHWC domain at this point in the graph then it means : 
This is a layout sensitive node which the EP initially claimed, but later after layout transformation it could no longer claim it. It is possible the bug is in EP (example EP checks for node domain and does not allow kMSNHWC domain). The reason I explicitly mentioned layout transformer because it is a good starting point for debugging.